### PR TITLE
Fix HttpWebRequest.KeepAlive handling

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/corefx/issues/18448
cc: @davidsh, @Priya91

### CODE REVIEW

**Comment by stephentoub** [comment]
The failures are due to WinHttp apparently always sending a Connection: Keep-Alive header (unless it's suppressed) and libcurl only doing so if explicitly asked to do so.

File: src/System.Net.Requests/src/System/Net/HttpWebRequest.cs [hunk 1]
```diff
@@ -384,24 +384,7 @@ public string TransferEncoding
...
-                {
-                    SetSpecialHeaders(HttpKnownHeaderNames.KeepAlive, bool.TrueString);
-                }
-                else
-                {
-                    SetSpecialHeaders(HttpKnownHeaderNames.KeepAlive, bool.FalseString);
-                }
-            }
-        } 
+        public bool KeepAlive { get; set; } = true;
```
Comment (davidsh) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: I don't understand this fix.

By the way, the KeepAlive property is a throwback to the early days of HTTP/1.0 where users needed to specify that they wanted to have the connection stay around.  Once HTTP/1.1 was done, the default semantics are for persistent connections and the connection keepalive request headers are by default sent. Thus, the defaults are usually for KeepAlive being true anyways.

Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: > I don't understand this fix.

What don't you understand?

We track whether KeepAlive is set, in a field, as is done on desktop.  Then when we create the request, we set the appropriate header based on KeepAlive.

Comment (davidsh) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: I don't understand this particular C# syntax:

```c#
public bool KeepAlive { get; set; } = true;
```

What does it mean to do the "= true;" part?

Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: Ah.  It means that it's an auto-prop that's initialized to true.  The compiler generates something like:
```C#
private bool _backingFieldKeepAlive = true;

public bool KeepAlive
{
    get { return _backingFieldKeepAlive; }
    set { _backingFieldKeepAlive = value; }
}
```

Comment (davidsh) [inline src/System.Net.Requests/src/System/Net/HttpWebRequest.cs]: ok.  thanks for clarifying.

**Pull request merged**